src/processor_fallback.cpp

```diff
@@ -0,0 +1,185 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+// Fallback processor detection and dispatch
+
+namespace Fallback {
+
+static inline const std::string &host_cpu_name()
+{
+    static std::string name = jl_get_cpu_name_llvm();
+    return name;
+}
+
+static const std::vector<TargetData<1>> &get_cmdline_targets(void)
+{
+    auto feature_cb = [] (const char*, size_t, FeatureList<1>&) {
+        return false;
+    };
+    return ::get_cmdline_targets<1>(feature_cb);
+}
+
+static std::vector<TargetData<1>> jit_targets;
+
+static TargetData<1> host_target_data(void)
+{
+    return TargetData<1>{host_cpu_name(), jl_get_cpu_features_llvm(), {{}, 0}, {{}, 0}, 0};
+}
+
+static TargetData<1> arg_target_data(const TargetData<1> &arg, bool require_host)
+{
+    TargetData<1> res = arg;
+    if (res.name == "native") {
+        res.name = host_cpu_name();
+        append_ext_features(res.ext_features, jl_get_cpu_features_llvm());
+    }
+    return res;
+}
+
+static uint32_t sysimg_init_cb(const void *id)
+{
```

Comment:
This function (and various others in this file), don't really seem to need to be duplicated and conditionally compiled.

Reply:
Most of these functions have more difference than the code they can share. Using a callback is certainly possible but that'll make the code much harder to read (so I only use it for a few functions that has a lot of shared logic like cmdline parsing and sysimg loading). A few internal API functions can be moved out but they are small enough that I think being able to find them in a file easily is more useful than reducing a few lines of code. Some of the API function implementations also need platform specific tweaking so either they can't be all moved out (making it hard to find code) or new conditional compilation need to be added in the function (which also decrease readability).

The feature comparison part in `sysimg_init_cb` is indeed one of the few good candidate to be separated out as a helper function though one of the most important condition `max_vector_size` needs to be a callback, making it quite a bit more ugly.... (And FWIW, it took a bit of an effort to twist the condition on different archs to all fit this code and I am still impressed it actually works out with the same code..............................)

Reply:
> Using a callback is certainly possible

This is C++, so you can also use interfaces or templates, if that helps.

Reply:
Yes, I know. But a large number of those makes the the code really hard to read......